geo::OpDetGeo made aware of detector orientation (issue #26118)

PetrilloAtWork opened the pull request:
Details of the philosophy, design and implementation of this request are documented in [Redmine issue #26118](https://cdcvs.fnal.gov/redmine/issues/26118).

PetrilloAtWork committed:
GeometryBuilderStandard can pass extra arguments to object constructors

PetrilloAtWork committed:
GeometryBuilderStandard code order moved

PetrilloAtWork committed:
Added default constructor to geo::PlaneBase

PetrilloAtWork committed:
OpDetGeo is provided with reference directions

PetrilloAtWork committed:
RealComparisons_test converted to BOOST_TEST

PetrilloAtWork committed:
Added parallel and orthogonal checks to Vector3DComparisons

PetrilloAtWork committed:
Documentation and minor maintenance on Geometry code

PetrilloAtWork committed:
Updated geo::OpDetGeo definitions, interface and implementation.

It now utilizes the information about the orientation.
This is discussed in Redmine issue #26118, which this commit is aimed to be the resolution of.

FNALbuild commented:
A new Pull Request was created by  @PetrilloAtWork (Gianluca Petrillo) for develop.

It involves the following packages:

larcorealg


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//978/console

knoepfel commented on larcorealg/Geometry/details/OpDetGeoSpecFillers.h:
```diff
@@ -0,0 +1,178 @@
+
+// C/C++ standard libraries
+#include <cmath>
+#endif // 0
```
Remove disabled code.

knoepfel replied:
Sorry, Gianluca.  Working on a review and prematurely submitted this.  The comment stands, though.

PetrilloAtWork replied:
Yup, I left something behind. Fixed.

knoepfel commented on larcorealg/Geometry/details/OpDetGeoSpecFillers.cxx:
```diff
@@ -0,0 +1,379 @@
+// Framework libraries
+#include "cetlib_except/exception.h"
+
+#endif // 0
```
Remove disabled code.

knoepfel replied:
Thanks.

knoepfel commented on test/CoreUtils/RealComparisons_test.cc:
```diff
@@ -19,100 +19,550 @@
+  constexpr lar::util::RealComparisons check(1e-5f);
 
-  double const sqrt2 = std::sqrt(2);
+  constexpr double const sqrt2 = std::sqrt(2);
```
`std::sqrt` is not a `constexpr` function, even though GCC provides `constexpr` implementation.  One should not assume this of (e.g.) Clang, though.

PetrilloAtWork replied:
I suspect this is not the first time I remove `constexpr` from there...
<rant>It's √2. √2. **_√2_**!! we should not need to compute it each time.</rant>
Removed.

knoepfel commented on larcorealg/CoreUtils/RealComparisons.h:
```diff
@@ -175,42 +181,94 @@ namespace lar {
+      template <typename VectA, typename VectB>
+      constexpr bool nonParallel(
+        VectA const& a, Value_t normSqA, VectB const& b, Value_t normSqB) const
+        { return !parallel(a, normSqA, b, normSqB); }
```
The `non*` functions add more maintenance without providing much improvement for the user.  Please remove them.

PetrilloAtWork replied:
Done...

knoepfel commented on larcorealg/CoreUtils/RealComparisons.h:
```diff
@@ -175,42 +181,94 @@ namespace lar {
+      /// Returns whether the specified vectors are not orthogonal.
+      template <typename VectA, typename VectB>
+      constexpr bool nonOrthogonal(VectA const& a, VectB const& b) const
+        { return !orthogonal(a, b); }
```
Please remove.

PetrilloAtWork replied:
... done...

knoepfel commented on larcorealg/CoreUtils/RealComparisons.h:
```diff
@@ -240,21 +298,76 @@ namespace lar {
+      template <typename VectA, typename VectB>
+      constexpr bool nonParallel(
+        VectA const& a, Value_t normSqA, VectB const& b, Value_t normSqB) const
+        { return !parallel(a, normSqA, b, normSqB); }
```
Please remove.

PetrilloAtWork replied:
... done...

knoepfel commented on larcorealg/CoreUtils/RealComparisons.h:
```diff
@@ -240,21 +298,76 @@ namespace lar {
+      /// Returns whether the specified vectors are not orthogonal.
+      template <typename VectA, typename VectB>
+      constexpr bool nonOrthogonal(VectA const& a, VectB const& b) const
+        { return !orthogonal(a, b); }
```
Please remove.

PetrilloAtWork replied:
... and done.

PetrilloAtWork committed:
Changed geo::OpDet closest TPC algorithm

PetrilloAtWork committed:
Removed stale code.

FNALbuild commented:
Pull request #21 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//979/console

PetrilloAtWork commented:
@knoepfel There was a `TODO` in the code, a small algorithm that I wanted to change. It's `computeMinDistSq()` implementation (`CryostatGeo.cxx`). I apologise if I make you go backward in reviewing...

knoepfel commented on larcorealg/Geometry/OpDetGeo.cxx:
```diff
@@ -24,23 +29,17 @@ namespace geo{
-
+    : fTrans{ makeTrans(std::move(trans)) }
+    , fOpDetNode{ &node }
+    , fShape{ makeShape(*fOpDetNode) }
```
Not to squelch your enthusiasm for C++17 😉, but mixing `std::variant`-based polymorphism with virtual-based polymorphism introduces more points of maintenance than it removes.  I think what is required is that the `fSpecs` object can be created/overwritten during the `UpdateAfterSorting` call.  Instead of using the `std::variant`, is it possible to bind to the correct specs-maker function in the constructor and then just call it where necessary?  For example:

```cpp
fSpecsMaker{ specsMakerFor(*fOpDetNodex) } // Creates specs-maker bounded to pointer of derived shape type
// ... and then in UpdateAfterSorting
fSpecs = fSpecsMaker(...);
```
I've not thought this through fully, so you may end up with something that's more complicated than what you have.

PetrilloAtWork replied:
I was indeed experimenting with `variant`...
There is a lot to say here.

First, I agree that mixing this `variant` "polymorphism" with ROOT's is awkward, and I would rather have only one of them used. I did not go all the way through, and at the conclusion of this discussion I will gladly change the class to adopt only one of them.

Then: the "traditional" polymorphism via inheritance has at this level gone stale: all the uses except one (in `geo::OpDetGeo::PrintOpDetInfo()`) are via `dynamic_cast`ing to the final type. Given the nature of `geo::OpDetGeo`, which shape is used is determined at construction and never changed: all these `dynamic_cast()` just rediscover over and over the same answer, and if anything make the code a tiny bit more obscure.
For this reason, a more sensible choice in my opinion is to store which one of the shapes is present, and then use that static information when needed. The `variant` object has exactly this information and a bit of interface around it, so it felt like a reasonable option rather than a less expressive integer or a more expressive but still less immediate enumerator.

Moreover, it has been my dark master plan (never carried to the end) since my first rehaul of `geo::OpdetGeo` (not this one: years ago) to detach from the underlying polymorphism and to attempt to give `geo::OpDetGeo()` a uniform interface. This may also be biasing my choice against dynamic casts. In my ideal world, there would be no access to `Shape()` or `GetNode()`. Admittedly those ones are still there and the whole thing may be a lost fight, since the performance of photodetectors depends _a lot_ on the details of their shape.

Last, I am not very fond of having a data member in the class that is used exactly once in the object lifetime, at ~construction~ initialisation. I think your approach can straightforwardly be moved to the `UpdateAfterSorting()` though, to initialise a local "spec maker" to be used and then disposed of.

Ok, after this maximally boring minimal essay, the real point of discussion is the second one: which approach is the most appropriate.
As you can guess, I lean toward the more static one, with `dynamic_cast()` confined in a small little corner as an unavoidable evil.

PetrilloAtWork replied:
(I moved the pull request to draft under the persuasion that any outcome of this discussion will require either a new implementation or some finalisation work before reaching merge quality)

PetrilloAtWork replied:
Note that I still need feedback on this point.

knoepfel replied:
Gianluca, I've been thinking about this for awhile, and I'm having trouble coming up with an elegant solution.  Let's go with what you have.

knoepfel reviewed (commented):
Thanks for the PR, Gianluca.  Looks pretty good in general.  Two main comments:

1. Please remove the (e.g.) `nonParallel` and `nonOrthogonal` negation functions--users have to wonder if they should be calling (e.g.) `nonParallel` instead of `not parallel`, which you don't want people to worry about.  Those "companion" functions provide little syntactic benefit at the expense of more code to maintain.
2. Generally `std::variant` and standard `virtual` polymorphism achieve different aims.  Mixing them is kind of awkward.  At the same time, ROOT has provided its own awkward interface for dealing with these shapes, so you may have selected a reasonable solution.  I provide one possible option that does not require using `std::variant`...but the solution you end up with may not be any better than what you already have.

I feel more strongly about (1) than (2), but I'm happy to follow up on anything.

PetrilloAtWork committed:
Removed negative vector comparisons from Vector#DComparison

As for code review recommendation.

PetrilloAtWork committed:
Removed non-standard constexpr from unit test.

Probably for the second time.

FNALbuild commented:
Pull request #21 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//985/console

PetrilloAtWork committed:
Merge branch 'develop' into feature/gp_Issue26118

FNALbuild commented:
Pull request #21 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//1034/console

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
-LArSoft tests failed on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12991&builds= 
 -build

lgarren commented:
@PetrilloAtWork  alas, there is a c7 build failure.  You may need to make sure this branch is up to date with the head of develop.

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12990&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5953&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12990&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6673&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12990&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/6208&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12990&builds=

FNALbuild commented:
+sbnd tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/7428&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12990&builds=

FNALbuild commented:
-dune tests failed, with ignored failure for unit_test, on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/12063&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12990&builds=

FNALbuild commented:
-icarus tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/2758&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12990&builds=

lgarren commented:
@PetrilloAtWork  Please review the dune unit and CI test failures.

PetrilloAtWork commented:
Yes, of course the branch is not up to date with `develop`. I'll take a look.

PetrilloAtWork committed:
Merge remote-tracking branch 'upstream/develop' into feature/gp_Issue26118

PetrilloAtWork committed:
Clang does not consider my friends as classy.

FNALbuild commented:
Pull request #21 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//1042/console

PetrilloAtWork commented:
My algorithm has some issue with the DUNE dual phase geometry. I need to understand and fix that.

PetrilloAtWork committed:
Fixed error message in geo::OpDetGeoSpecFillers

PetrilloAtWork committed:
geo::OpDetGeo is now asked to point to the center of the closest TPC

FNALbuild commented:
Pull request #21 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//1043/console

PetrilloAtWork committed:
Restored geo::OpDetGeo::GetCenter() functionality pre-sorting

Experiment sorting code may need it.

FNALbuild commented:
Pull request #21 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//1044/console

PetrilloAtWork commented:
I added a feature (make PMT point toward the centre of the closest TPC), which appears to fix ProtoDUNE Dual Phase geometry issue.
I have also fixed a bug (`GetCenter()` was not guaranteed before sorting, but sorting code may well decide to use it — ICARUS did, but unit tests would have caught it only if they were in `debug` mode).

One unit test from DUNE still fails (`prodsingle_dune35t` does not fine the module `SimCounter35t`), but this does not seem a failure related to the changes in this pull request.

Ready for second review and second test...

knoepfel commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
-LArSoft tests failed on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13039&builds= 
 -build

lgarren commented:
@PetrilloAtWork  See the c7 build failure

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13038&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5965&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13038&builds=

FNALbuild commented:
-argoneut tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/6218&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13038&builds=

FNALbuild commented:
-sbnd tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/7466&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13038&builds=

FNALbuild commented:
-dune tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/12080&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13038&builds=

FNALbuild commented:
-uboone tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6684&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13038&builds=

lgarren commented:
@PetrilloAtWork  The CI tests are impacted right now because dcache is down for maintenance.  But please see https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13039&builds=

FNALbuild commented:
-icarus tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/2776&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13038&builds=

lgarren commented:
@PetrilloAtWork c7 build failure:
[ 45%] Building CXX object larcorealg/larcorealg/Geometry/CMakeFiles/larcorealg_Geometry.dir/details/OpDetGeoSpecFillers.cxx.o
/home/garren/scratch/larsoft/v09_34_00/srcs/larcorealg/larcorealg/Geometry/details/OpDetGeoSpecFillers.cxx:362:26: error: 
      calling a protected constructor of class 'geo::details::OpDetGeoSpecFillerFromTube'
  { assert(node); return OpDetGeoSpecFillerFromTube{ trans, dirs }(*node); }
                         ^
/home/garren/scratch/larsoft/v09_34_00/srcs/larcorealg/larcorealg/Geometry/details/OpDetGeoSpecFillers.h:68:33: note: 
      declared protected here
  using OpDetGeoSpecFillerBase::OpDetGeoSpecFillerBase; // inherit constructor
                                ^
/home/garren/scratch/larsoft/v09_34_00/srcs/larcorealg/larcorealg/Geometry/details/OpDetGeoSpecFillers.cxx:365:26: error: 
      calling a protected constructor of class 'geo::details::OpDetGeoSpecFillerFromSphere'
  { assert(node); return OpDetGeoSpecFillerFromSphere{ trans, dirs }(*node); }
                         ^
/home/garren/scratch/larsoft/v09_34_00/srcs/larcorealg/larcorealg/Geometry/details/OpDetGeoSpecFillers.h:85:33: note: 
      declared protected here
  using OpDetGeoSpecFillerBase::OpDetGeoSpecFillerBase; // inherit constructor
                                ^
/home/garren/scratch/larsoft/v09_34_00/srcs/larcorealg/larcorealg/Geometry/details/OpDetGeoSpecFillers.cxx:368:26: error: 
      calling a protected constructor of class 'geo::details::OpDetGeoSpecFillerFromBox'
  { assert(node); return OpDetGeoSpecFillerFromBox{ trans, dirs }(*node); }
                         ^
/home/garren/scratch/larsoft/v09_34_00/srcs/larcorealg/larcorealg/Geometry/details/OpDetGeoSpecFillers.h:114:33: note: 
      declared protected here
  using OpDetGeoSpecFillerBase::OpDetGeoSpecFillerBase; // inherit constructor
                                ^
3 errors generated.

knoepfel commented:
@PetrilloAtWork, see https://github.com/LArSoft/larcorealg/pull/21#issuecomment-948032984.

PetrilloAtWork committed:
Fiixes for Clang.

FNALbuild commented:
Pull request #21 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//1047/console

PetrilloAtWork commented:
I am puzzled why GCC did not catch that "issue" (C++ legalese, but still).
Anyway, I have fixed (apparently, "fiixed"), compiled and tested the branch with Clang 7 (`c7:prof`).

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13101&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13102&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5982&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13102&builds=

FNALbuild commented:
-uboone tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6706&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13102&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/6235&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13102&builds=

FNALbuild commented:
-icarus tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/2811&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13102&builds=

FNALbuild commented:
-sbnd tests warning, with build warning,, with ignored warning for build, on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/7538&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13102&builds=

FNALbuild commented:
-dune tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/12106&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13102&builds=

lgarren commented:
I *think* the CI tests are normal.  The icarus failure is worrying, but seems unrelated to this PR.

pgreen135 commented:
The changes in ci_g4_regression_test_dunefd are a bit worrying to me, it looks like the amount of light seen has reduced due to this PR (dunefd is running the light simulation mode that makes use of these functions). Although sbnd which is using the same method is unaffected... I can investigate tomorrow

PetrilloAtWork commented:
@pgreen135 if confirmed that is _definitely_ worrying.
FYI: I'll be away for a couple of weeks, so I don't guarantee I will be able to address the issue shortly.

pgreen135 commented:
@PetrilloAtWork this is indeed causing an issue for dunefd-hd (the dune10kt_v4_1x2x6.gdml geometry) . The optical detector sensitive windows in this case are defined as: 
`<box name="ArapucaAcceptanceWindow" lunit="cm"
      x="2.2"
      y="9.3"
      z="46.8"/>`

But, in larsim I am now seeing `opDet.Length() = 2.2, opDet.Height() = 9.3`.  Whereas previously Length() would return 46.8, hence the large reduction in the light picked up by the CIs.

These are not rotated at all relative to the TPC, so as far as I can tell this shouldn't be happening with the new functionality:
`<rotation name="rIdentity"            unit="deg" x="0" y="0"   z="0"/>` 
`<physvol>
       <volumeref ref="volOpDetSensitive_0-0-0"/>
       <position name="posOpArapuca0-0-0-TPC-12-0-0" unit="cm" 
         x="-0.0500000000000002" 
       y="-612.31875" 
         z="-535.175"/>
       <rotationref ref="rIdentity"/>
     </physvol>
`

pgreen135 commented:
I can also confirm this same issue is present for SBND @PetrilloAtWork; its a bit concerning this wasn't immediately obvious from the CI tests but that's a whole other issue...

PetrilloAtWork commented:
Many thanks, @pgreen135.
Conveniently, I can't log in the GPVM from where I am, so it will likely be a couple of weeks before I can take a look to this.
In the meanwhile, I am moving this request to draft.

lgarren commented:
@PetrilloAtWork  When you are able to get back to this, you may need to start a new PR from the head of develop.

knoepfel commented:
@PetrilloAtWork, this PR has been dormant for a while.  Can we close this?  We are currently re-evaluating the geometry design.  If you'd like to create a new PR, please talk with us so we can coordinate.
